engines/ospray/ispc/render/SimulationMaterial.cpp

```diff
@@ -0,0 +1,74 @@
+/* Copyright (c) 2015-2018, EPFL/Blue Brain Project
+ * All rights reserved. Do not distribute without permission.
+ * Responsible Author: Juan Hernando <juan.hernando@epfl.ch>
+ *
+ * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "SimulationMaterial.h"
+#include "SimulationMaterial_ispc.h"
+
+#include "../geometry/Cones.h"
+#include "../geometry/SDFGeometries.h"
+
+#include <brayns/common/geometry/Cone.h>
+#include <brayns/common/geometry/Cylinder.h>
+#include <brayns/common/geometry/SDFGeometry.h>
+#include <brayns/common/geometry/Sphere.h>
+#include <brayns/common/geometry/Sphere.h>
+
+#include <ospray/SDK/geometry/Cylinders.h>
+#include <ospray/SDK/geometry/Geometry.h>
+#include <ospray/SDK/geometry/Spheres.h>
+
+#include <map>
+#include <typeindex>
+#include <typeinfo>
+
+extern "C" {
+int SimulationMaterial_getBytesPerPrimitive(const void* geometry)
+{
+    const ospray::Geometry* base =
```

Comment:
How often is this called? I am concerned about the performance of doing several dynamic casts like this.

Reply:
It's called every shaded hit. The dynamic cast is not ideal but I started implementing another alternative using std::type_index that I though it would be faster, but it won't as it requires hashing of strings, something which is for sure slower than a dynamic_cast. I haven't seen any noticeable performance difference in the experiments I've done with a debug build.

In any case, I'd go for correctness first and optimizations later in a separate PR.

Reply:
A potentially faster solution would be to keep track of all geometry instances in a std::unordered_map<void*, size_t> that stores what this function should return. Hashing a pointer and doing a lookup on this table seems faster than RTTI. However, the code to keep track of the instances will be a lot dirtier.

Reply:
Is it possible to store an enum in the user data of the geometry?

Reply:
Hahaha, is that a joke?

There's no such a think as Geometry user data, I wished it existed, otherwise I wouldn't have spent so much time with ispc generating stupid code for something it shouldn't accept as correct code in the first place.

Reply:
Sadly, it is not a joke. I am confused by this since you do get the geometry data pointer like this:

```
   // The data pointer in all "derived" geometries is just after data members
    // of the base Geometry struct.
    const uniform uint8* data =  *((const uniform uint8**)&geometry[1]);
```

So could you not put the enum first then?

Reply:
No, because the Geometry struct is created by ISPC code in OSPRay. There's no way I can change or tinker with that. My code is already type unsafe as it's making an assumption about the location of this data pointer, so don't think I'm happy with it, but I didn't find a better way without modifying OSPRay or going back to have our own spheres and cylinders.

Reply:
Okay, we leave it like this then.